clang/include/clang/Sema/Sema.h

```diff
@@ -3165,7 +3165,16 @@ class Sema final {
 
   /// Diagnose any unused parameters in the given sequence of
   /// ParmVarDecl pointers.
-  void DiagnoseUnusedParameters(ArrayRef<ParmVarDecl *> Parameters);
+  ///
+  /// Normally, we check if the parameter decls have the Referenced bit set.
+  /// C++ Coroutines, however, are a special case due to the existences of
+  /// parameter moves (See Sema::buildCoroutineParameterMoves), the parameters
+  /// are always referenced in coroutines. Therefore, in the case of coroutines,
+  /// CoroutineBodyRefs must be passed to correctly diagnose parameter usages
+  /// as written by the user.
+  void DiagnoseUnusedParameters(
+      ArrayRef<ParmVarDecl *> Parameters,
+      llvm::SmallSet<ParmVarDecl *, 4> *CoroutineBodyRefs = nullptr);
```

Comment:
Let's try to not polluting the interfaces as much as possible. Since most of the functions are not coroutines, it looks not good to pollute the interfaces.

Personally, I prefer the style:

```
if (the function is coroutine)
     DiagnoseUnusedParametersInCoroutines(...);
else
     DiagnoseUnusedParameters(...); // Original call.
```

Also it may be better to use `FunctionScopeInfo::CoroutineParameterMoves` than yet another DeclsVisitor. It is easy prone and hard to maintain.

Reply:
If so, what about adding an assertion in the original version of `DiagnoseUnusedParameters`? Or should we keep the interface as is (allowing coroutines to be diagnosed in the same function, but handle the special case in `DiagnoseUnusedParameters` without changing the interface?)

`CoroutineParameterMoves` does not give me what I want. It points me to `Stmt`s that moved the parameters while I need the opposite -- All other references to the parameters.

Reply:
> If so, what about adding an assertion in the original version of DiagnoseUnusedParameters? Or should we keep the interface as is (allowing coroutines to be diagnosed in the same function, but handle the special case in DiagnoseUnusedParameters without changing the interface?)

Sorry, what do you mean by adding an assertion to DiagnoseUnusedParameters? Do you say doing so after adopting my suggest? Then, it sounds good but how can we do that? And it sounds not meaningful.

> CoroutineParameterMoves does not give me what I want. It points me to Stmts that moved the parameters while I need the opposite -- All other references to the parameters.

Sorry for misleading. But let's try to get a solution light-weight enough.  I don't want to pay a lot for this diagnostic .  An instinct idea now is to mark all the parameters as non referenced/used after building CoroutineParameterMoves. But it needs to be verified.

Reply:
> Sorry, what do you mean by adding an assertion to DiagnoseUnusedParameters? Do you say doing so after adopting my suggest? Then, it sounds good but how can we do that? And it sounds not meaningful.

I didn't make it clear. I was asking what you may think about an assertion in DiagnoseUnusedParameters that !CurFunction()->isCoroutine()

> An instinct idea now is to mark all the parameters as non referenced/used after building CoroutineParameterMoves. But it needs to be verified.

New to clang so I can't tell if I did something wrong. But the compiler crashed when I initially tried that. 

I have to sign off now. Catch up with you soon.